src/protocols/secure_channel/CASESession.cpp

```diff
@@ -884,12 +880,116 @@ CHIP_ERROR CASESession::SendSigma1()
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorSessionId)), input.initiatorSessionId));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kDestinationId)), input.destinationId));
+
+    VerifyOrReturnError(input.pEphPubKey != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(tlvWriter.PutBytes(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorPubKey)), *input.pEphPubKey,
+                                            static_cast<uint32_t>(input.pEphPubKey->Length())));
+
+    VerifyOrReturnError(input.initiatorMrpConfig != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    ReturnErrorOnFailure(
+        EncodeSessionParameters(TLV::ContextTag(ToRaw(Sigma1Tags::kInitiatorMRPParams)), *input.initiatorMrpConfig, tlvWriter));
+
+    if (input.sessionResumptionRequested)
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kResumptionID)), input.resumptionId));
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(ToRaw(Sigma1Tags::kResume1MIC)), input.initiatorResumeMICSpan));
+    }
+
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&msg));
+
+    return CHIP_NO_ERROR;
+}
+
 CHIP_ERROR CASESession::HandleSigma1_and_SendSigma2(System::PacketBufferHandle && msg)
 {
     MATTER_TRACE_SCOPE("HandleSigma1_and_SendSigma2", "CASESession");
-    ReturnErrorOnFailure(HandleSigma1(std::move(msg)));
 
-    return CHIP_NO_ERROR;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Parse and Validate Received Sigma1, and decide next step
+    SuccessOrExit(err = HandleSigma1(std::move(msg)));
+
+    switch (mNextStep)
+    {
+    case Step::kSendSigma2: {
+
+        System::PacketBufferHandle msg_R2;
+        EncodeSigma2Inputs encodeSigma2;
+
+        // TODO verify MATTER_LOG_METRIC locations
+        MATTER_LOG_METRIC_BEGIN(kMetricDeviceCASESessionSigma2);
+
+        SuccessOrExit(err = PrepareSigma2(encodeSigma2));
+        SuccessOrExit(err = EncodeSigma2(msg_R2, encodeSigma2));
+        SuccessOrExitAction(err = SendSigma2(msg_R2), MATTER_LOG_METRIC_END(kMetricDeviceCASESessionSigma2, err));
+
+        mDelegate->OnSessionEstablishmentStarted();
+        break;
+    }
+    case Step::kSendSigma2Resume: {
+
+        System::PacketBufferHandle msg_R2_resume;
+        EncodeSigma2ResInputs encodeSigma2Resume;
+
+        MATTER_LOG_METRIC_BEGIN(kMetricDeviceCASESessionSigma2Resume);
+
+        SuccessOrExit(err = PrepareSigma2Resume(encodeSigma2Resume));
+        SuccessOrExit(err = EncodeSigma2Resume(msg_R2_resume, encodeSigma2Resume));
+        SuccessOrExitAction(err = SendSigma2Resume(msg_R2_resume),
+                            MATTER_LOG_METRIC_END(kMetricDeviceCASESessionSigma2Resume, err));
+
+        mDelegate->OnSessionEstablishmentStarted();
+        break;
+    }
+        // TODO should I keep this?
+    case Step::kSendStatusReport:
```

Comment:
Yes. If kSendStatusReport is the outcome from Sigma1, it means there is a status report to send, and we need to send it:

```

        // FindLocalNodeFromDestinationId returns CHIP_ERROR_KEY_NOT_FOUND if validation of DestinationID fails, which will trigger
        // status Report with ProtocolCode = NoSharedTrustRoots.
        mNextStep = Step::kSendStatusReport;
```

So I recommend moving the following block in kSendStatusReport case:

```
    if (err == CHIP_ERROR_KEY_NOT_FOUND)
    {
        SendStatusReport(mExchangeCtxt, kProtocolCodeNoSharedRoot);
        mState = State::kInitialized;
    }
```

Reply:
Actually I made the `case Step::kSendStatusReport` automatically redirect us to exit  through `ExitNow()` and kept the `SendStatusReport` with `NoSharedTrustRoots` in `exit` because:
- Technically, it is possible to get a `CHIP_ERROR_KEY_NOT_FOUND` in other Code Paths than the one you quoted (where I defined the NextStep); such as through calls to `SignWithOpKeypair`, called in `PrepareSigma2()`. 
- I don't know if getting to that codepath is impossible or not, so I thought I'll just keep that `SendStatusReport` in `exit` and redirect that `case` to exit (which does make the `NextStep` kind of useless.

- What do you think?

Reply:
I have changed the Structure in a way to make `HandleSigma1` return a Variant with either Step OR CHIP_ERROR which will trigger a status report anyways; with the Switch case only handling the "Sigma Sending" Part.